### PR TITLE
Toggle assertions generate doc

### PR DESCRIPTION
**Pull request by rouson**
* Add two new assertions to verify conformable remote puts of coarray arrays.
* Add FORD documentation generator file [doc-generator.md](./doc-generator.md).
* Enhance and update the README.md with CMake build instructions and [FORD] documentation generation instructions.

[FORD]: https://github.com/cmacmackin/ford

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -29,6 +29,11 @@ if ("${CMAKE_CURRENT_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_BINARY_DIR}")
     )
 endif()
 
+#Toggle C preprocessor macro for turning assertions on or off
+if(NO_ASSERTIONS)
+  add_definitions(-DNO_ASSERTIONS)
```
Comment (gutmann) [inline CMakeLists.txt]: Could make this `-DUSE_ASSERTIONS=\".False.\"`

File: src/utilities/assertions_interface.F90 [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  !!
+  implicit none
+  private
+  public :: assert
+  public :: assertions
+  
+! Conditioning assertion calls on this compile-time constant enables optimizing compilers
+! to eliminate assertion calls during a dead-code removal phase of optimization
+#ifdef NO_ASSERTIONS
+  logical, parameter :: assertions=.false.
```
Comment (gutmann) [inline src/utilities/assertions_interface.F90]: And this section just 
`assertions = USE_ASSERTIONS`

File: CMakeLists.txt (hunk 1 above)
Comment (gutmann) [reply CMakeLists.txt]: (I'm not sure about the \" part...)

File: src/utilities/assertions_interface.F90 (hunk 2 above)
Comment (rouson) [reply src/utilities/assertions_interface.F90]: Nice!  I'll try this out and let you know how it goes.

**Pull request merged**